src/R/Components/Impl/PackageManager/Implementation/RestartManager.cs

```diff
@@ -0,0 +1,143 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Security;
+
+namespace Microsoft.R.Components.PackageManager.Implementation {
+    internal class RestartManager {
+        /// <summary>
+        /// Given a list of filenames with aboslute path, returns list of process currently locking those files.
+        /// </summary>
+        /// <param name="filePaths">Filenames with absolute path.</param>
+        /// <returns>List of processes locking the files. null if it encounters any error.</returns>
+        public static IList<Process> GetProcessesUsingFiles(string[] filePaths) {
+            uint sessionHandle;
+            List<Process> processes = new List<Process>();
+
+            int restartManager = NativeMethods.RmStartSession(out sessionHandle, 0, Guid.NewGuid().ToString("N"));
+            if (restartManager != 0) {
+                return null;
+            }
+
+            try {
+                restartManager = NativeMethods.RmRegisterResources(sessionHandle, (uint)filePaths.Length, filePaths, 0, null, 0, null);
+                if (restartManager != 0) {
+                    return null;
+                }
+
+                uint pnProcInfoNeeded = 0, pnProcInfo = 0, lpdwRebootReasons = RmRebootReasonNone;
+                restartManager = NativeMethods.RmGetList(sessionHandle, out pnProcInfoNeeded, ref pnProcInfo, null, ref lpdwRebootReasons);
+
+                if (restartManager == ERROR_MORE_DATA) {
+                    RM_PROCESS_INFO[] processInfo = new RM_PROCESS_INFO[pnProcInfoNeeded];
+                    pnProcInfo = (uint)processInfo.Length;
+
+                    restartManager = NativeMethods.RmGetList(sessionHandle, out pnProcInfoNeeded, ref pnProcInfo, processInfo, ref lpdwRebootReasons);
```

Comment:
Again, assuming the list can change between consecutive `RmGetList` calls, this might end up filling fewer than `pnProcInfoNeeded` items in the array - I'm assuming that `pnProcInfo` is `ref` precisely so that the function can communicate the actual number of items it has filled. The loop below should take that into account.